feat: Add support for proper exception handling

kavirajk opened the pull request:
## Summary
<!-- A short description of the changes with a link to an open issue. -->

Related: #1468 

This PR adds support to extract the exception happens during mid-stream. Supports generic exception marker for both CH 24.11 and before.


## Checklist
Delete items not relevant to your PR:
- [x] Unit and integration tests covering the common scenarios were added
- [ ] A human-readable description of the changes was provided to include in CHANGELOG
- [ ] For significant changes, documentation in https://github.com/ClickHouse/clickhouse-docs was updated with further explanations or tutorials

kavirajk committed:
feat: Add support for proper exception handling

Signed-off-by: Kaviraj <kavirajkanagaraj@gmail.com>

kavirajk committed:
fix without zero-chunk at the end

Signed-off-by: Kaviraj <kavirajkanagaraj@gmail.com>

kavirajk committed:
add tests for http exception handling

Signed-off-by: Kaviraj <kavirajkanagaraj@gmail.com>

kavirajk committed:
use proper test connection

Signed-off-by: Kaviraj <kavirajkanagaraj@gmail.com>

chernser commented on conn_http.go:
```diff
@@ -433,12 +433,130 @@ func (h *httpConnect) readData(reader *chproto.Reader, timezone *time.Location)
+		// "__exception__" marker as binary data
+
+		// Try to read any remaining data
+		lr := &limitedReader{reader: reader, limit: 100 * 1024}
```
This relates to the block size. What if block size is 1Mb and exception is at the end?

chernser commented on conn_http.go:
```diff
@@ -433,12 +433,130 @@ func (h *httpConnect) readData(reader *chproto.Reader, timezone *time.Location)
+	dataStr := string(data)
+
+	// Verify exception marker exists
+	if !strings.Contains(dataStr, "__exception__") {
```
this is redundant because next `if` checks it again using proper way.

kavirajk replied:
good catch. Removed it.

chernser commented on conn_http.go:
```diff
@@ -433,12 +433,130 @@ func (h *httpConnect) readData(reader *chproto.Reader, timezone *time.Location)
+	}
+
+	// Skip past first __exception__\r\n
+	pos := firstMarker + len("__exception__")
```
we can save `len("__exception__")` for further use.

kavirajk committed:
PR remarks
1. Move magic numbers to well defined constants
2. Fix the block size when decoding

Signed-off-by: Kaviraj <kavirajkanagaraj@gmail.com>

kavirajk committed:
Merge branch 'main' into kavirajk/exception-tag-support
